### PR TITLE
Add dist hooks support for custom device

### LINKED ISSUE #104389: distributed hooks want to support custom device
### 🐛 Describe the bug

for distributed hooks, there are some hard-codes, for cutsom device (privateuse1 backend, cuda-like device), we also want to support these hooks.

### Versions

main

### PR DESCRIPTION
**Pull request by dilililiwhy**
Fixes #104389 

Try to add dist hooks support for custom device again due to original PR https://github.com/pytorch/pytorch/pull/104155 is marked with stale label.

original desc

> 1. Now for distributed hooks, there are some hard-code like cuda, we want to support these hooks for custom device( privateuse1 backend), so we use the abstract device-module to run some funcs. 
> 2. In torch/nn/parallel/distributed.py, I want to define a variable self.device_module = getattr(torch, self.device_type, None) so that we can reuse it, but it will cause an error in serialization, TypeError: cannot pickle 'module' object. So we call the func by getattr when needs.

 Function _get_device_module is landed in this PR https://github.com/pytorch/pytorch/pull/107289

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @kiukchung @lucasllc

### CODE REVIEW

**Comment by albanD** [comment]
cc @wconstab who's the right person to review DDP PRs?

**Comment by dilililiwhy** [comment]
cc @fduwjj @rohan-varma @wz337 
Could you take a look at this PR？

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -694,7 +694,11 @@ def __init__(
             )
 
         self.device_type = next(iter(distinct_device_types))
-
+        # use `torch.cuda` module to solve `xla` device
+        self._device_module_type = (
+            self.device_type if self.device_type != "xla" else "cuda"
```
Comment (wconstab) [inline torch/nn/parallel/distributed.py:699]: Can you explain this part
Cc @alanwaketan

File: torch/distributed/utils.py [hunk 2]
```diff
@@ -337,3 +338,11 @@ def _replace_by_prefix(
 
 def _data_ptr_allocated(tensor: torch.Tensor) -> bool:
     return tensor.untyped_storage().data_ptr() > 0
+
+
+class _DeviceModule(Protocol):  # noqa: PYI046
```
Comment (wconstab) [inline torch/distributed/utils.py:343]: Are you using the protocol here to represent a python module (e.g. torch.cuda is a module that defines these methods)?  It makes sense I think, but I've never seen this technique, is it a recommended/pythonic way to do things?

Comment (dilililiwhy) [reply torch/distributed/utils.py:343]: Yes, protocol here is used for device module static check, as other devices may not inherited torch.cuda gernerally. I think it is a formal way in type hinting (PEP484 and PEP544)

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (dilililiwhy) [reply torch/nn/parallel/distributed.py:699]: As torch.xla is not existed, this part added for some xla tests in CI flow while it seems that xla reuses cuda-related code in distributed.py.

Comment (dilililiwhy) [reply torch/nn/parallel/distributed.py:699]: Maybe I can remove this part, and try to find the actual issue.

Comment (dilililiwhy) [reply torch/nn/parallel/distributed.py]: @wconstab 
Could you help to trigger the ciflow?

**Comment by dilililiwhy** [comment]
@albanD
Sorry to bother. How to acquire a CI approval for this PR due to some cases cannot be verified in local.

**Comment by albanD** [comment]
@wconstab can do that when he reviews the PR!

**Comment by dilililiwhy** [comment]
> @pytorchbot label "ciflow/trunk"

It seems that adding label directly is not allowed. Maybe there is an **Approve and run** button for CIFlow?

**Review by wconstab — COMMENTED** [review]
id like to get some other input on the best UX to expose here. I think its functionally OK.  Would be nice if users didn't have to call 2 lines of code to get the device type, then get the module, then finally call module.stream or module.synchronize().

Any ideas @kwen2501 @wanchaol @shuqiangzhang

**Pull request closed without merging**